### ISSUE TITLE
FR: GitHub integration

### ISSUE THREAD

**Issue by dbarnett** [martinvonz/jj#4555]
Should have some [forge integrations](https://github.com/martinvonz/jj/blob/HEAD/docs/roadmap.md#forge-integrations) for GitHub remotes (avoiding the need to rely on https://cli.github.com/ and colocated git repos).

Some possible use cases could be:

1. creating/managing PRs
2. interacting with `issues` or code (`browse`) related to the current repo and its state
3. checking status on remote action workflows

**Comment by PhilipMetzger** [comment]
> Some possible use cases could be:
> 
> 1. creating/managing PRs
> 2. interacting with `issues` or code (`browse`) related to the current repo and its state
> 3. checking status on remote action workflows

That sounds almost like the functionality which Phabricator's Arcanist (`arc`) had, although it never was completed. If we want Forge integrations as a first class thing, then we'd probably need to cover these.

**Comment by dbarnett** [comment]
Yeah, tbh all the integrations I can think of for GitHub would have direct analogs for other forges, and I'm questioning whether it needs anything like an explicit `jj github` command at all. Besides some slight [terminology differences](https://about.gitlab.com/blog/2017/09/11/comparing-confusing-terms-in-github-bitbucket-and-gitlab/), if you can do it for GH you should probably be able to do it for any other supported remote.

The gotchas making it all generic would be: 

1. mechanism to track which forge(s) you have associated with which git remotes and topics might be cumbersome
2. discoverability and troubleshooting might be slightly harder due to terminology differences 

BTW I saw relevant comments in another issue about whether to support multiple remotes in the same `push` command… this would have some of the same challenges at another level, if your repo has remotes on multiple forges.

**Comment by PhilipMetzger** [comment]
> Yeah, tbh all the integrations I can think of for GitHub would have direct analogs for other forges, and I'm questioning whether it needs anything like an explicit `jj github` command at all.

Tbh, 1.5 years ago, I'd agreed with you, but I had this discussion a few times in the Discord but Martin had a [good point here][dc1], which changed my opinion. 

> The gotchas making it all generic would be:
> 
> 1. mechanism to track which forge(s) you have associated with which git remotes and topics might be cumbersome
> 2. discoverability and troubleshooting might be slightly harder due to terminology differences

Imo, it is worthwhile to have the common generic implementation in the `lib`, while each Forge is responsible for its own idiosyncrasies in their trait implementation.

> BTW I saw relevant comments in another issue about whether to support multiple remotes in the same `push` command… this would have some of the same challenges at another level, if your repo has remotes on multiple forges.

Agreed, as there is no _canonical main repository_ without user intervention. 

[dc1]: https://discord.com/channels/968932220549103686/969291218347524238/1128718254576255076

**Comment by steveklabnik** [comment]
I think there's two concerns here: sharing code, and the UX. If the features are similar, there's no reason they couldn't share code. But I think there are UX benefits to forge specific names: it's a very clear indicator of support of your particular forge, it's easy to find, and there will always be some features that are specific to some forges that could get awkward otherwise.

Then again, I'm also speaking of "forges" as maybe a bit broader of a topic than they are. Is Gerrit a forge? It's really just a code review tool. What about sourcehut? It supports many features that github/gitlab/bitbucket do, but also has things they don't, like real time chat, and uses mailing lists instead of PRs/changesets.

**Comment by arxanas** [comment]
The discussion in this thread regards forges in general, so I continued the discussion here: https://github.com/martinvonz/jj/issues/485#issuecomment-2386594529

**Comment by dbarnett** [comment]
Excellent, thanks for pointing me toward the context! 

So IIUC there's no concern if I send a PR to implement e.g. a `github|gh pr create` command?

**Comment by PhilipMetzger** [comment]
> So IIUC there's no concern if I send a PR to implement e.g. a `github|gh pr create` command?

There never was, although I'd prefer to keep it under `github` as `gh` conflicts with their own CLI. Maybe take a look at the `gerrit send` PR (#2845) for some inspiration and things to share.

**Comment by dbarnett** [comment]
Thanks!

> as `gh` conflicts with their own CLI.

What would that affect? You're saying `jj gh pr create` and commands like that could be somehow ambiguous because of the similarity to their CLI name?

**Comment by PhilipMetzger** [comment]
> What would that affect? You're saying `jj gh pr create` and commands like that could be somehow ambiguous because of the similarity to their CLI name?

Names bring user expectations which could imply some false dependency on the `gh` tool itself. The `github` namespace also makes sense to contrast `gerrit`, which lands with `gerrit send`, but this is again just my opinion,

**Comment by martinvonz** [comment]
FWIW, we already have `jj git` and a bunch of subcommands that conflict with with `git` and its subcommands. I'm not arguing for `jj gh` (I think I also prefer `jj github`, at least the canonical form).

**Comment by mati865** [comment]
Using a full service name is the safest option, if somebody wants a shorter name, they should create an alias.

**Comment by dbarnett** [comment]
AFAICT the logistics for this would be something like...

1. Add a dependency on https://crates.io/crates/octorust
2. Create a GitHub App to use for auth flow from a command like `jj github auth login` 
3. Somehow integrate the client secret securely into the jj app

For 2, octorust has [instructions for authenticating](https://docs.rs/octorust/0.7.0/octorust/#authenticating-github-apps) and I created https://github.com/apps/jujutsu-vcs but I don't really know what I'm doing on permissions and key security. If anyone wants me to transfer that (or delete it so they can recreate fresh), LMK.

Alternatively it could just shell out to a separately-installed gh cli, which would be less "self-contained" but I believe would avoid some of the security complexities and allow seamlessly reusing existing tokens a user may have already set up with gh cli.

**Comment by arxanas** [comment]
FWIW I shelled out to `gh` in git-branchless's GitHub integration specifically for those reasons. Authentication is a lot of incidental complexity whose effort could be better-spent elsewhere. I believe you can query the API directly via `gh` as well, in case you wanted to issue some esoteric queries.

Whatever solution should probably also handle Enterprise GitHub + single-sign-on as well (which I assume `gh` does).

**Comment by dbarnett** [comment]
> Whatever solution should probably also handle Enterprise GitHub + single-sign-on as well (which I assume gh does).

Yep, looks like `$GH_HOST` and `$GH_ENTERPRISE_TOKEN` are the [env variables](https://cli.github.com/manual/gh_help_environment) for controlling that, although I might need help from someone who regularly uses Enterprise to verify that's all fully working as intended.

From what I can tell, all the gh commands do have mechanisms to hard-code the repo/branch/etc to target, although they seem to vary a bit per command and might be a little brittle:

| jj cmd | gh cmd |
| --------- | ----------- |
| `jj github pr view` | `GH_REPO=someorg/somerepo gh pr view BRANCHNAME` |
| `jj github browse some/path` | `GH_REPO=someorg/somerepo gh browse -b BRANCHNAME some/path` |

where I believe `GH_REPO` can be inferred from git remote URLs (and probably same for `GH_HOST`?).

Those `pr` and `browse` ones jump out at me as some good initial ones to support, plus probably some `auth` commands so usage isn't too clunky if you're not freshly authenticated. The ones that have no relationship to "current repo" (like `gist` commands or `repo list`) I don't see any value in supporting as jj commands. Others like `run`/`secret`/`workflow` I could imagine adding support for as we see demand.

Maybe someday particular extensions like `gh act` too, but that doesn't seem to respect `GH_REPO` and I'm not sure there'd be a non-brittle way to generally support extensions like that if they remain needing special-case overrides.

**Comment by martinvonz** [comment]
The commands I would find most useful would be for:

* Create/update stack of PRs, setting the base branch between them
* Fetch the commit from a PR without having to configure a remote

**Comment by PhilipMetzger** [comment]
While I would've preferred to build our own thing, @arxanas point still stands that we'll never get to the point of being a first-class user to GitHub, which `gh` is. So going with `gh` should be fine, even if it feels a bit icky to me. 

For other providers (#485), we hopefully should be able to have a first-class experience built-in. 

I think the option to use a stacked workflow or the normal GitHub branch-like thing would also be good to have in the initial version.

**Comment by martinvonz** [comment]
For creating/updating a stack of PRs, what you can't do in a forge-agnostic way is to set the base branch, so we'd need that to be under `jj github`.

For fetching, the only thing I was thinking of that might help if it had GithHub-awareness is so you can specify a PR number instead of a ref.

**Comment by emilazy** [comment]
Note that there’s two options for fetching GitHub PRs, the HEAD commit (`pulls/…/head`) or the synthesized merge commit (`pulls/…/merge`).

I would love a stacked PR solution that lets you choose which commits go into which PRs rather than making every PR a single commit, but there might not be many people who want that.

**Comment by dbarnett** [comment]
> I would love a stacked PR solution that lets you choose which commits go into which PRs rather than making every PR a single commit

Yep, that's exactly what "topics" will accomplish, isn't it? Each topic would generally map 1:1 with a PR?

**Comment by emilazy** [comment]
Probably! I don’t really fully understand topics except that they will solve everything and make life perfect forever, so that sounds plausible.

**Comment by PhilipMetzger** [comment]
> Yep, that's exactly what "topics" will accomplish, isn't it? Each topic would generally map 1:1 with a PR?

Yes, if the topic strictly maps to a topological branch (which then translates into a Git branch). Although having the option to not strictly enforce that they must not overlap would still be nice, since it allows a multi-head approach to sending a PR.

**Comment by arxanas** [comment]
> Yes, if the topic strictly maps to a topological branch (which then translates into a Git branch). 

[topic: topics] To expand on this (seems relevant to GitHub interop), a "topic" for our purposes is some subset of commits. A branch, and therefore pull request, can only be inferred if that set has a single head. (You can make the conditions more restrictive as desired.)

[topic: commit message tagging] I've seen many people establish topics/branches by convention in the commit message specifically for tooling purposes.
- Example: [git-branchstack](https://github.com/krobelus/git-branchstack) is used for many-to-many topic management.
- It may work for many people in practice to use the commit message to establish the associated logical topic.
  - For example, an organization may have a policy to put a link to a bug tracker in the commit message, which could be used to determine the associated topic as well.
  - If implementing this, it could use a user-defined regex to determine the topic. (I don't think it's worth doing in a first pass.)

[topic: git-branchless design] git-branchless implicitly addresses this in its GitHub forge at present by only creating branches for the commits in the provided revset.
- Example invocation: `git submit --forge github --create 'stack() & message(JIRA-123)'`.
- Note: in git-branchless, creating PRs is an explicit step. When invoking without `--create`, only commits in the provided revset which already have PRs are updated. jj could make a different design decision here.
- Ancestor commits not part of the revset implicitly go into the head commit PR.
  - There is no explicit implementation — that's just how it works if you create and push a branch and create a PR.
  - But that _is_ the deliberate design for git-branchless to handle multi-commit PRs.

**Comment by arxanas** [comment]
If anyone starts to implement stacked PRs for the GitHub forge, here's my report on relevant issues to consider: https://github.com/arxanas/git-branchless/discussions/1259

**Comment by dbarnett** [comment]
And still seems like topics will take some time yet to implement, so initial GH integration might not include any of that in its scope. I expected for now to create a PR from a bookmark name and then include all commits from base to that bookmark in the PR, same as the gh tool does.

**Comment by martinvonz** [comment]
That sounds good to me, fwiw. (I can't speak for others, of course.)

**Comment by parthsdoshi** [comment]
@dbarnett Have you already taken a stab at
> I expected for now to create a PR from a bookmark name and then include all commits from base to that bookmark in the PR, same as the gh tool does.

If so, it would be nice if I could see any artifacts you may have. I've been a user of `jj` for some time now and thought I'd try and help out depending on complexity... not promising anything given my novice abilities 

~Lastly, I see the word `topic` used above -- is there a design document or another issue I can read through to understand what they are and if they've been implemented already?~
Nvm, found it: https://github.com/jj-vcs/jj/pull/4496

**Comment by calebdw** [comment]
I came here looking for this because the current workflow to edit someone else's PR is clunky, it would be really nice if this could be streamlined.

In the past (as a maintainer of a repo), all I had to do was `gh pr checkout <number>`, make the changes, and then `git push`---no need to setup a remote or track a branch: the submitter's PR would be updated and I could merge it.


Using jj, this is much more convoluted: after running `gh pr checkout <number>`, I then have to go find their github repo fork, copy the git ssh url, add it as a remote in my local repository, fetch so I have the refs, and then track the branch with that remote. After I merge the PR I then have to remove their fork from my remotes list because I don't care about it anymore.

**Comment by PhilipMetzger** [comment]
> I came here looking for this because the current workflow to edit someone else's PR is clunky, it would be really nice if this could be streamlined.
> 
> In the past (as a maintainer of a repo), all I had to do was `gh pr checkout <number>`, make the changes, and then `git push`---no need to setup a remote or track a branch: the submitter's PR would be updated and I could merge it.
> 
> Using jj, this is much more convoluted: after running `gh pr checkout <number>`, I then have to go find their github repo fork, copy the git ssh url, add it as a remote in my local repository, fetch so I have the refs, and then track the branch with that remote. After I merge the PR I then have to remove their fork from my remotes list because I don't care about it anymore.

While I get your pain this is not the issue you want for this, as the core of this problem is described in this issue https://github.com/jj-vcs/jj/issues/4388 which is what `gh` does under the hood anyway.

**Comment by petr-tik** [comment]
GitHub support for stacked PRs is in beta. Would be awesome to integrate with it

https://github.github.com/gh-stack/

**Comment by Munksgaard** [comment]
> GitHub support for stacked PRs is in beta. Would be awesome to integrate with it
> 
> https://github.github.com/gh-stack/

Already addressed [here](https://github.com/jj-vcs/jj/issues/485#issuecomment-4244264055)

**Comment by ilyapoz** [comment]
> avoiding the need to rely on https://cli.github.com/ and colocated git repos

I want to question this a bit. How strong of a requirement is it to avoid colocated git repos? I mean, to work with github without a colocated working copy?

Using gh cli to avoid authentication issues I believe is agreed upon.

upd: ok, I see that this is probably not an issue is we rely on the same code as jj git to deliver the code to the remote

**Comment by ilyapoz** [comment]
heads up: I'm trying my hand on getting this off the ground. I want to set a goal of "do an MVP that does at least one useful function"

I'm using these docs for reference: https://github.com/jj-vcs/jj/pull/7483
https://github.com/jj-vcs/jj/pull/9498

I like the 'bookmarkless' approach from the first one. I like "open source vs proprietary" separation from the second one. I will focus on proprietary first, because this is my use case and because AFAIK there is no way to create a PR from fork into main repo without contributor permissions. There is a ticket about something like that in gh-stack: https://github.com/github/gh-stack/issues/46

What do you say in general terms about the minimal bar required to have this in main? I mean feature-wise?
For example (say stop when enough, but plz say as soon as possible)
1. jj github submit that allows at least to submit a single PR with `main` as base
2. jj github submit that uses `gh-stack` to submit a stack with a simple "one change per PR" semantic
3. jj github status that shows some useful info about existing PRs in "current stack" (for some definition of current stack)

I would really like to have only 1 in main just to work in reasonable bites. Maybe as a feature turned off by default in release?

**Comment by ilyapoz** [comment]
I have researched a bit more into https://github.com/jennings/jj-spr

I believe that feature wise it is quite interesting and relevant. For example, the "force only fast forward updates on the forge by creating artificial commits" is quite good "workaround for GH forge's limitations". I would not push this as "must have feature", but maybe a nice to have as an option. 

On the other hand, this project tries to work around the fact that GH does not have native stacking support (jj-spr has a special command `land` that changes branches while merging the first PR in the stack). I don't think I want to implement this and target the `gh-stack` eventually once it comes out of beta and for now keep all this in the hands of users. For example, right now with "squash" merge strategy + merge queue enabled the second PR in the stack gets boggled, because github changes the base branch to main and commits from the first PR to merge get into the next one. This should be fixed with gh-stack. This is the commit that demonstrates how git-spr solves this: https://github.com/spacedentist/spr/pull/26 they actually create a synthetic "pseudo-stack": instead of master <- A <- B <- C They create master <- A   master.A <- B master.B <- C: these `master.<>` branches are just there to 'pin base branches' for github to show the correct diff and in this case if A is merged, then there is no such problem with commits from A getting into the review for B. I still think I don't want to get all these hacks into the new integration and just wait for gh-stack to be released. In the meanwhile we can just support basic things without stacks.

**Comment by ilyapoz** [comment]
ok, this project now completely demotivated me https://github.com/mrjones2014/jj-gh :)
